File: sample-conference-videochat/sample-conference-videochat/Components/Authorization/Classes/InputContainer/InputContainer.m
//
//  InputContainer.m
//  sample-conference-videochat
//
//  Created by Injoit on 01.12.2020.
//  Copyright © 2020 QuickBlox. All rights reserved.
//

#import "InputContainer.h"
#import "UITextField+Chat.h"
#import "UIColor+Chat.h"
#import "NSString+Chat.h"

static const CGFloat padding = 12.0f;
static const CGFloat cornerRadius = 4.0f;
NSString *const defaultColor = @"#DFEBFF";
NSString *const activeColor = @"#ACBFE2";

@interface InputContainer()
//MARK: - IBOutlets
@property (weak, nonatomic) IBOutlet UILabel *titleLabel;
@property (weak, nonatomic) IBOutlet UITextField *inputTextfield;
@property (weak, nonatomic) IBOutlet UILabel *hintLabel;
//MARK: - Properties
@property (assign, nonatomic) BOOL valid;
@property (strong, nonatomic) NSArray<NSString *> *regexes;
@property (strong, nonatomic) NSString *hint;
@end

@implementation InputContainer
// MARK: - Public Methods
- (NSString *)text {
    return self.inputTextfield.text;
}

- (void)setInputEnabled:(BOOL)inputEnabled {
    [self.inputTextfield setEnabled:inputEnabled];
}

- (void)inputTextfieldBecomeFirstResponder {
    [self.inputTextfield becomeFirstResponder];
}

- (void)setupWithTitle:(NSString *)title hint:(NSString *)hint regexes:(NSArray<NSString *> *)regexes {
    [self.inputTextfield addShadow:[UIColor colorWithHexString:defaultColor] cornerRadius:cornerRadius];
    [self.inputTextfield setPadding:padding isLeft:YES];
    self.titleLabel.text = title;
    self.hint = hint;
    self.hintLabel.text = @"";
    self.regexes = regexes;
    self.valid = NO;
}

- (BOOL)isValid {
    return self.valid;
}

//MARK: - Internal Methods
- (void)setValid:(BOOL)valid {
    _valid = valid;
    if ([self.delegate respondsToSelector:@selector(inputContainer:didChangeValidState:)]) {
        [self.delegate inputContainer:self didChangeValidState:valid];
    }
    if ((self.inputTextfield.text.length == 0 && self.inputTextfield.isFirstResponder == NO) || valid == YES) {
        self.hintLabel.text = @"";
    } else {
        self.hintLabel.text = self.hint;
    }
}

- (IBAction)editingDidBegin:(UITextField *)sender {
    [self validateTextField:sender];
    [sender addShadow:[UIColor colorWithHexString:activeColor] cornerRadius:cornerRadius];
}
- (IBAction)editingChanged:(UITextField *)sender {
    [self validateTextField:sender];
}
- (IBAction)editingDidEnd:(UITextField *)sender {
    [self validateTextField:sender];
    [sender addShadow:[UIColor colorWithHexString:defaultColor] cornerRadius:cornerRadius];
}

- (void)validateTextField:(UITextField *)texField {
    if (!texField.text) {
        self.valid = NO;
        return;
    }
    self.valid = [texField.text validateWithRegexes:self.regexes];
}

@end

File: sample-chat-swift/sample-chat-swift/Components/Authorization/InputContainer/InputContainer.swift
//
//  InputContainer.swift
//  sample-chat-swift
//
//  Created by Injoit on 25.11.2020.
//  Copyright © 2020 QuickBlox. All rights reserved.
//

import UIKit

struct InputContainerConstant {
    static let padding: CGFloat = 12.0
    static let cornerRadius: CGFloat = 4.0
}

protocol InputContainerDelegate: AnyObject {
    func inputContainer(_ container: InputContainer, didChangeValidState isValid: Bool)
}

class InputContainer: UIView {
    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var inputTextfield: UITextField!
    @IBOutlet weak var hintLabel: UILabel!
    weak var delegate: InputContainerDelegate?
    var hint = "Set a hint for your textfield here"
    var regexes:[String] = []
    
    private(set) var valid: Bool = false {
        didSet {
            delegate?.inputContainer(self, didChangeValidState: valid)
            if (inputTextfield.text?.isEmpty == true && inputTextfield.isFirstResponder == false) || valid == true {
                hintLabel.text = ""
            } else {
                hintLabel.text = hint
            }
        }
    }
    
    func setup(title: Title, hint: Hint, regexes: [Regex]) {
        inputTextfield.setPadding(left: InputContainerConstant.padding)
        inputTextfield.addShadow(color: #colorLiteral(red: 0.8755381703, green: 0.9203008413, blue: 1, alpha: 1), cornerRadius: InputContainerConstant.cornerRadius)
        titleLabel.text = title.rawValue
        self.hint = hint.rawValue
        hintLabel.text = ""
        self.regexes = regexes.compactMap { $0.rawValue }
    }
    
    @IBAction func editingDidBegin(_ sender: UITextField) {
        validate(texField: sender)
        sender.addShadow(color: #colorLiteral(red: 0.6745098039, green: 0.7490196078, blue: 0.8862745098, alpha: 1), cornerRadius: InputContainerConstant.cornerRadius)
    }
    @IBAction func editingChanged(_ sender: UITextField) {
        validate(texField: sender)
    }
    @IBAction func editingDidEnd(_ sender: UITextField) {
        validate(texField: sender)
        sender.addShadow(color: #colorLiteral(red: 0.8745098039, green: 0.9215686275, blue: 1, alpha: 1), cornerRadius: InputContainerConstant.cornerRadius)
    }
    
    private func validate(texField: UITextField) {
        guard let text = texField.text else {
            valid = false
            return
        }
        valid = text.isValid(regexes: regexes)
    }
}

extension String {
    func isValid(regexes: [String]) -> Bool {
        for regex in regexes {
            let predicate = NSPredicate(format: "SELF MATCHES %@", regex)
            if predicate.evaluate(with: self) == true {
                return true
            }
        }
        return false
    }
}
